### PR TITLE
pkg/endpoint: attempt to restore failed endpoints regenerations from previous life

### PR DESCRIPTION
**Pull request by aanm**
Signed-off-by: André Martins <andre@cilium.io>

Fixes: https://github.com/cilium/cilium/issues/4769

```release-note
Attempt to restore failed endpoint regenerations from a previous cilium-agent execution
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5182)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

File: daemon/state.go [hunk 1]
```diff
@@ -267,13 +268,13 @@ func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) error {
 
 // readEPsFromDirNames returns a list of endpoints from a list of directory
```
Comment (ianvernon) [inline daemon/state.go]: Update this function's documentation to reflect that it's a mapping of endpoint ID to endpoint, not just a list of endpoints.

File: daemon/state.go [hunk 2]
```diff
@@ -316,7 +317,15 @@ func readEPsFromDirNames(basePath string, eptsDirNames []string) []*endpoint.End
 			scopedLog.WithError(err).Warn("Unable to parse the C header file")
 			continue
 		}
-		possibleEPs = append(possibleEPs, ep)
+		if _, ok := possibleEPs[ep.ID]; ok {
+			// If the endpoint already exists then give priority to the directory
+			// that contains and endpoint that didn't fail to be build.
```
Comment (ianvernon) [inline daemon/state.go]: Nit: and --> an

File: daemon/state_test.go [hunk 3]
```diff
@@ -140,15 +140,43 @@ func (ds *DaemonSuite) generateEPs(baseDir string, epsWanted []*e.Endpoint, epsM
...
+			}
+			epsNames = append(epsNames, ep.FailedDirectoryPath())
+
+			// create one failed and the other non failed directory for ep 256.
+			if ep.ID == 256 {
+				fullDirName := filepath.Join(baseDir, ep.DirectoryPath())
+				os.MkdirAll(fullDirName, 777)
+
+				ep.UnconditionalLock()
+				// Change endpoint a little bit to make sure we didn't
```
Comment (ianvernon) [inline daemon/state_test.go]: Can you elaborate on this? How exactly does doing this ensure that we do not "give priority to the failed endpoint regeneration"?

Comment (aanm) [reply daemon/state_test.go]: The test is not here, it's below. I rephrase this a little bit.

File: daemon/state.go (hunk 2 above)
Comment (aanm) [reply daemon/state.go]: Done

File: daemon/state.go (hunk 1 above)
Comment (aanm) [reply daemon/state.go]: Done

**Comment by aanm** [comment]
test-me-please

File: pkg/endpoint/endpoint.go [hunk 4]
```diff
@@ -1219,11 +1219,14 @@ func (e *Endpoint) GetIdentity() identityPkg.NumericIdentity {
...
-func (e *Endpoint) directoryPath() string {
+// DirectoryPath returns the directory name for this endpoint bpf program.
+func (e *Endpoint) DirectoryPath() string {
 	return filepath.Join(".", fmt.Sprintf("%d", e.ID))
 }
 
-func (e *Endpoint) failedDirectoryPath() string {
+// FailedDirectoryPath returns the directory name for this endpoint bpf program
+// failed builds.
+func (e *Endpoint) FailedDirectoryPath() string {
```
Comment (raybejjani) [inline pkg/endpoint/endpoint.go:1229]: Does this now need a read-lock? :(
e.ID never changes so maybe we can just say that and leave a comment on e.ID saying don't ever change it or something (it feels a little silly to lock on this)

Comment (raybejjani) [reply pkg/endpoint/endpoint.go:1229]: oh, I see `mutex` is below `ID` in the class definition. I think our style means that it isn't guarded by the mutex?

File: pkg/endpoint/endpoint.go [hunk 5]
```diff
@@ -1360,7 +1363,8 @@ func FilterEPDir(dirFiles []os.FileInfo) []string {
 	eptsID := []string{}
 	for _, file := range dirFiles {
 		if file.IsDir() {
-			if _, err := strconv.ParseUint(file.Name(), 10, 16); err == nil {
+			_, err := strconv.ParseUint(file.Name(), 10, 16)
+			if err == nil || strings.HasSuffix(file.Name(), "_next_fail") {
```
Comment (raybejjani) [inline pkg/endpoint/endpoint.go:1367]: This is fine but would a regex be more clear? and capture all the variations more clearly? shrug. I'm not sure if it's worth it.

File: pkg/endpoint/endpoint.go [hunk 6]
```diff
@@ -1360,7 +1363,8 @@ func FilterEPDir(dirFiles []os.FileInfo) []string {
 	eptsID := []string{}
```
Comment (raybejjani) [inline pkg/endpoint/endpoint.go:1363]: It's worth mentioning in the function comment that it now also returns `_next_failed` directories as well.

File: daemon/state.go [hunk 7]
```diff
@@ -265,15 +266,15 @@ func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) error {
...
 
-// readEPsFromDirNames returns a list of endpoints from a list of directory
-// names that can possible contain an endpoint.
-func readEPsFromDirNames(basePath string, eptsDirNames []string) []*endpoint.Endpoint {
-	possibleEPs := []*endpoint.Endpoint{}
-	for _, epID := range eptsDirNames {
-		epDir := filepath.Join(basePath, epID)
+// readEPsFromDirNames returns a mapping of endpoint ID to endpoint of endpoints
+// from a list of directory names that can possible contain an endpoint.
+func readEPsFromDirNames(basePath string, eptsDirNames []string) map[uint16]*endpoint.Endpoint {
```
Comment (raybejjani) [inline daemon/state.go:271]: It would be good to explain that this also returns failed endpoints.

**Review by raybejjani — APPROVED** [review]
a more general question: do we need to worry about starting controllers?
I think the CEP sync ones are started when added to endpointmanager, so that's ok. `resolve-identity-%d`, `sync-policymap-``sync-identity-to-k8s-pod` ,`sync-%s-identity-mapping` get started in various ways, usually related to changing labels, regenerating etc.

Since we're now, as a new thing, restoring not-ready endpoints, we might end up with a different set of running controllers for the endpoint compared to if it was a normally created EP that then failed a regeneration.

**Comment by tgraf** [comment]
@aanm @raybejjani Did you figure out whether you want to address @raybejjani's feedback or not?

**Comment by raybejjani** [comment]
Nothing I said should be blocking, and it can be fixed later. It was mostly comment things (the mutex bit is probably what I said in the follow up).

**Pull request merged**